README.md
### CSG Pro Website

[![Commitizen friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/) [![Build Status](https://travis-ci.org/csgpro/csgpro.com.svg?branch=master)](https://travis-ci.org/csgpro/csgpro.com)
[![NSP Status](https://nodesecurity.io/orgs/csg-pro/projects/fa8a9d6e-9137-4db1-8245-82547db77603/badge)](https://nodesecurity.io/orgs/csg-pro/projects/fa8a9d6e-9137-4db1-8245-82547db77603)

### Running Locally
To run the application you need to have node `v6.7.0`.

The database needs to be restored and you'll need to create a `settings.json` file in the root of the project. See [settings.sample.json](./settings.sample.json).

### Prerequisites
1. Node v6.7.0 or higher
2. Restore database backup
3. Rename `sample.env` to `.env` & set appropriate values

Open Terminal & Run:  
`npm start`

Then open your browser to `http://localhost:3000`


CHANGELOG.md
<a name="3.4.2"></a>
# [3.4.1](https://github.com/csgpro/csgpro.com/compare/v3.4.1...v3.4.2) (2017-09-25)


### Bug Fixes

* Update Address

<a name="3.4.1"></a>
# [3.4.1](https://github.com/csgpro/csgpro.com/compare/v3.4.0...v3.4.1) (2017-05-10)


### Bug Fixes

* legacy routes to properly redirect blog posts ([fe68baa](https://github.com/csgpro/csgpro.com/commit/fe68baa))



<a name="3.4.0"></a>
# [3.4.0](https://github.com/csgpro/csgpro.com/compare/v3.3.0...v3.4.0) (2017-04-17)


### Bug Fixes

* **robots.txt:** disable security to expose robots.txt file ([7911ffc](https://github.com/csgpro/csgpro.com/commit/7911ffc)), closes [#185](https://github.com/csgpro/csgpro.com/issues/185)


### Features

* **Twitter Card:** Add twitter card meta info ([000fd1d](https://github.com/csgpro/csgpro.com/commit/000fd1d))



<a name="3.3.0"></a>
# [3.3.0](https://github.com/csgpro/csgpro.com/compare/v3.2.1...v3.3.0) (2017-04-06)


### Features

* **Logger:** Add logger ([44d4fe6](https://github.com/csgpro/csgpro.com/commit/44d4fe6))
* **Privacy Policy:** Adds privacy policy ([717be74](https://github.com/csgpro/csgpro.com/commit/717be74))

### Content Updates

* **Business Analytics:** Update Power BI Examples

<a name="3.2.1"></a>
## [3.2.1](https://github.com/csgpro/csgpro.com/compare/v3.2.0...v3.2.1) (2016-12-14)


### Bug Fixes

* **Admin Posts:** Fix category toggle. ([7796e06](https://github.com/csgpro/csgpro.com/commit/7796e06))
* **Post Api:** Include associations. ([6748593](https://github.com/csgpro/csgpro.com/commit/6748593))
* **Webhook:** Remove webhook commands and associated hooks. ([bacb9a8](https://github.com/csgpro/csgpro.com/commit/bacb9a8))



<a name="3.2.0"></a>
# [3.2.0](https://github.com/csgpro/csgpro.com/compare/v3.1.1...v3.2.0) (2016-11-30)


### Features

* **Post:** Add ability to delete posts (admin only) ([89345d2](https://github.com/csgpro/csgpro.com/commit/89345d2))
* **Post:** Add ability to filter by published/unpublished in admin ([2d43436](https://github.com/csgpro/csgpro.com/commit/2d43436))
* **Post:** Hide publish/un-publish/slug fields for non-admins ([8f1e0f9](https://github.com/csgpro/csgpro.com/commit/8f1e0f9))
* **Post:** prevent non-admins from changing post author. ([1d868e9](https://github.com/csgpro/csgpro.com/commit/1d868e9))
* **Post:** Set current user as author on new posts ([0a017df](https://github.com/csgpro/csgpro.com/commit/0a017df))
* **Post:** Show topics in list. Sort by post id. ([c2bcfcc](https://github.com/csgpro/csgpro.com/commit/c2bcfcc))



<a name="3.1.1"></a>
## [3.1.1](https://github.com/csgpro/csgpro.com/compare/v3.1.0...v3.1.1) (2016-11-12)


### Bug Fixes

* **Slugify:** Fix slugs. Add tests. ([ecfafea](https://github.com/csgpro/csgpro.com/commit/ecfafea))


### Features

* **Topics:** Add new topic when composing post ([db81f51](https://github.com/csgpro/csgpro.com/commit/db81f51))



<a name="3.1.0"></a>
# [3.1.0](https://github.com/csgpro/csgpro.com/compare/v3.0.2...v3.1.0) (2016-11-01)

* **Admin:** Upgrade to Angular 2.1
* **Admin:** Use Angular Material Module

<a name="3.0.2"></a>
## [3.0.2](https://github.com/csgpro/csgpro.com/compare/v3.0.1...v3.0.2) (2016-10-28)


### Bug Fixes

* **Topics:** return promise from getTopic call ([a5383f4](https://github.com/csgpro/csgpro.com/commit/a5383f4)), closes [#166](https://github.com/csgpro/csgpro.com/issues/166)



<a name="3.0.1"></a>
## [3.0.1](https://github.com/csgpro/csgpro.com/compare/v3.0.0...v3.0.1) (2016-10-23)


### Bug Fixes

* **Side Nav:** Restore links in side nav ([876319b](https://github.com/csgpro/csgpro.com/commit/876319b))



<a name="3.0.0"></a>
# [3.0.0](https://github.com/csgpro/csgpro.com/compare/v2.1.2...v3.0.0) (2016-10-14)



<a name="2.1.2"></a>
## [2.1.2](https://github.com/csgpro/csgpro.com/compare/v2.1.1...v2.1.2) (2016-10-07)



<a name="2.1.1"></a>
## [2.1.1](https://github.com/csgpro/csgpro.com/compare/v2.1.0...v2.1.1) (2016-09-03)


### Features

* **About:** Added seasonal header images ([024e6de](https://github.com/csgpro/csgpro.com/commit/024e6de))



<a name="2.1.0"></a>
# [2.1.0](https://github.com/csgpro/csgpro.com/compare/v2.0.0...v2.1.0) (2016-09-01)


### Bug Fixes

* View routes + deployment ([40b0674](https://github.com/csgpro/csgpro.com/commit/40b0674))
* **About:** Hide images in the sticky block. They force the sticky block width. ([2588958](https://github.com/csgpro/csgpro.com/commit/2588958)), closes [#149](https://github.com/csgpro/csgpro.com/issues/149)
* **Admin:** Fix missing authentication on post and put ([b1e859e](https://github.com/csgpro/csgpro.com/commit/b1e859e))
* **Admin:** Update login page to use Router Directives. ([39cd705](https://github.com/csgpro/csgpro.com/commit/39cd705))
* **Authorization:** Expire tokens after 30 days. Auto-renew token in admin app. ([871524d](https://github.com/csgpro/csgpro.com/commit/871524d))
* **build:** Fix migration, seeder paths. Fix copy-files to handle extensions. ([c4f3012](https://github.com/csgpro/csgpro.com/commit/c4f3012))
* **careers:** Fix link ([9611c9c](https://github.com/csgpro/csgpro.com/commit/9611c9c))
* **Code Highlighting:** Fix unkown code blocks from appearing blank. ([ab4cf80](https://github.com/csgpro/csgpro.com/commit/ab4cf80)), closes [#103](https://github.com/csgpro/csgpro.com/issues/103)
* **commitizen:** Restore cz-conventional-changelog for commitizen ([3694e74](https://github.com/csgpro/csgpro.com/commit/3694e74))
* **Contact Form:** Fix input type for company. Set to 'text'. ([9d9b5da](https://github.com/csgpro/csgpro.com/commit/9d9b5da)), closes [#97](https://github.com/csgpro/csgpro.com/issues/97)
* **Contact Form:** Fix phone number validation. Fix reset. Fix modal state. ([82f717a](https://github.com/csgpro/csgpro.com/commit/82f717a)), closes [#98](https://github.com/csgpro/csgpro.com/issues/98) [#100](https://github.com/csgpro/csgpro.com/issues/100)

app.js
/*jslint
  node: true*/

'use strict';

/**********************************
 * MODULE DEPENDENCIES
 **********************************/

var express          = require('express')
  , fs               = require('fs')
  , index            = require('./routes/index')
  , landing          = require('./routes/landing')
  , register         = require('./routes/register')
  , http             = require('http')
  , path             = require('path')
  , c                = require('nconf')
  , db               = require('./modules/db.js')
  , db2              = require('./modules/db2.js')
  , contact          = require('./routes/contact')
  , redirects        = require('./routes/redirects')
  , post             = require('./routes/post')
  , jobs             = require('./routes/jobs')
  , api              = require('./routes/api')
  , upload           = require('./routes/upload')
  , account          = require('./routes/account')
  , admin            = require('./routes/admin-post')
  , adminMain        = require('./routes/admin')
  , adminTopic       = require('./routes/admin-topic')
  , adminReminders   = require('./routes/admin-reminders')
  , passport         = require('passport')
  , TwitterStrategy  = require('passport-twitter').Strategy
  , LiveStrategy     = require('passport-windowslive').Strategy
  , jwt              = require('jwt-simple')
  , async            = require('async')
  , moment           = require('moment')
  , qs               = require('querystring')
  , request          = require('request')
  , app = module.exports.app = exports.app = express()
  , port             = 3000;


// Load the configuration file with our keys in it, first from the env variables
// then from the config.json file
c.env().file({ file: 'config.json' });

/**********************************
 * AUTHENTICATION
 **********************************/
var TWITTER_CONSUMER_KEY       = c.get('TWITTER_CONSUMER_KEY')
  , TWITTER_CONSUMER_SECRET    = c.get('TWITTER_CONSUMER_SECRET')
  , TWITTER_CALLBACK_URL       = c.get('TWITTER_CALLBACK_URL')
  , WINDOWS_LIVE_CLIENT_ID     = c.get('WINDOWS_LIVE_CLIENT_ID')
  , WINDOWS_LIVE_CLIENT_SECRET = c.get('WINDOWS_LIVE_CLIENT_SECRET')
  , WINDOWS_LIVE_CALLBACK_URL  = c.get('WINDOWS_LIVE_CALLBACK_URL')
  , SESSION_SECRET             = c.get('SESSION_SECRET');

/**********************************
 * SETTINGS / MIDDLEWARE
 **********************************/
app.set('views', __dirname + '/views');
app.set('view engine', 'jade');
app.use(express.logger());
app.use(express.compress());
app.use(express.cookieParser());
app.use(express.bodyParser());
app.use(express.methodOverride());
app.use(express.session({ secret: SESSION_SECRET }));
app.use(passport.initialize());
app.use(passport.session());

/**************
 * Robots.txt redirect
 **************/
app.use(function(req, res, next){
  var onAzure = req.host.indexOf("azurewebsites") > -1;
  if(req.url === '/robots.txt' && onAzure){
    req.url = '/robots-azure.txt';
  }
  next();
});
app.use(express.static(path.join(__dirname, 'public')));
app.use(app.router);

app.use(function(err, req, res, next) {
  res.render('error', { message : err });
});

/**************
 * Login Required Middleware
 **************/
function ensureAuthenticated(req, res, next) {
    if (!req.headers.authorization) {
        return res.status(401).send({ message: 'Please make sure your request has an Authorization header' });
    }

    var token = req.headers.authorization.split(' ')[1];
    var payload = jwt.decode(token, SESSION_SECRET);

    if (payload.exp <= moment().unix()) {
        return res.status(401).send({ message: 'Token has expired' });
    }

    req.user = payload.sub;
    next();
}

/****************
* JWT TIME
***************/
function createToken(user) {
    if(user.id) {
        var payload = {
            sub: user.id,
            iat: moment().unix(),
            exp: moment().add(14, 'days').unix(),
            user: user
        };
        return jwt.encode(payload, SESSION_SECRET);
    } else {
        return null;
    }
}

/**********************************
 * ROUTES
 **********************************/
app.get('/', index.homepage);
app.get('/power-bi/register', register.powerbi);

app.get('/admin/login', function (req, res) {
    res.redirect('/admin');
});

app.get('/post', post.getPostsBySearch);
app.get('/post/category/:category', post.getPostsByCategory);
app.get('/post/topic/:topic', post.getPostsByTopic);
app.get('/post/:id', post.getPostByID);

app.get('/jobs', jobs.getJobs);
app.get('/jobs/:id', jobs.getJobByID);

app.post('/contact', contact.index);
app.post('/csv', register.csv);

/*****************
 * ADMIN (deprecated)
 ****************/
app.get('/admin-old', auth, adminMain.index);
app.get('/admin-old/login', adminMain.login);

app.get('/admin-old/account'             , authAdmin , account.index);
app.post('/admin-old/account'            , authAdmin , account.create);
app.get('/admin-old/account/new'         , authAdmin , account.entry);
app.get('/admin-old/account/:id/update'  , authAdmin , account.update);
app.post('/admin-old/account/:id/update' , authAdmin , account.patch);
app.get('/admin-old/account/:id/delete'  , authAdmin , account.del);

app.get('/admin-old/topic'               , authAdmin , adminTopic.index);
app.post('/admin-old/topic'              , authAdmin , adminTopic.create);
app.get('/admin-old/topic/new'           , authAdmin , adminTopic.entry);
app.get('/admin-old/topic/:id/update'    , authAdmin , adminTopic.update);
app.post('/admin-old/topic/:id/update'   , authAdmin , adminTopic.patch);
app.get('/admin-old/topic/:id/delete'    , authAdmin , adminTopic.del);

app.get('/admin-old/post'                , auth      , admin.index);
app.get('/admin-old/post/new'            , auth      , admin.entry);
app.get('/admin-old/post/:id/update'     , auth      , admin.update);
app.post('/admin-old/post/:id/update'    , auth      , admin.create);
app.post('/admin-old/post'               , auth      , admin.create);
app.get('/admin-old/post/:id'            , auth      , admin.get);
app.get('/admin-old/posts'               , auth      , admin.all);
app.get('/admin-old/post/:id/publish'    , authAdmin , admin.publish);
app.get('/admin-old/post/:id/unpublish'  , authAdmin , admin.unpublish);
app.get('/admin-old/post/:id/delete'     , authAdmin , admin.del);

app.post('/admin-old/image-upload'       , auth      , admin.imageUpload);

app.get('/admin-old/notadmin', function(req, res) {
  res.send('You must be an admin to do the thing you were trying to do.');
});

app.get('/admin-old/reminders', authAdmin, adminReminders.index);
app.post('/admin-old/reminders', authAdmin, adminReminders.send);

/*****************
 * API
 ****************/

/* POSTS */
app.get('/api/posts', api.posts.getPosts);
app.get('/api/posts/category/:category', api.posts.getPosts);
app.get('/api/posts/topic/:topic', api.posts.getPosts);
app.get('/api/posts/:id', api.posts.getPostByID);
app.post('/api/posts', api.posts.createPost);
app.patch('/api/posts/:id', api.posts.updatePost);
app.delete('/api/posts/:id', api.posts.deletePost);

/* USERS */
app.get('/api/users', ensureAuthenticated, api.users.getUsers);
app.get('/api/users/:id', ensureAuthenticated, api.users.getUserByID);
app.get('/api/users/other/:search', api.users.getUserByOther);
app.post('/api/users', ensureAuthenticated, api.users.createUser);
app.patch('/api/users/:id', ensureAuthenticated, api.users.updateUser);
app.delete('/api/users/:id', ensureAuthenticated, api.users.deleteUser);

/* TOPICS */
app.get('/api/topics', api.topics.getTopics);
app.get('/api/topics/:id', api.topics.getTopicByID);
app.post('/api/topics', api.topics.createTopic);
app.patch('/api/topics/:id', api.topics.updateTopic);
app.delete('/api/topics/:id', api.topics.deleteTopic);

/* JOBS */
app.get('/api/jobs', api.jobs.getJobs);
app.get('/api/jobs/:id', api.jobs.getJobByID);
app.get('/api/jobs/:filter', api.jobs.getJobsByFilter);

/*****************
 * TWITTER
 ****************/
app.get('/auth/twitter', function(req, res) {
    var requestTokenUrl = 'https://api.twitter.com/oauth/request_token';
    var accessTokenUrl = 'https://api.twitter.com/oauth/access_token';
    var authenticateUrl = 'https://api.twitter.com/oauth/authenticate';

    if (!req.query.oauth_token || !req.query.oauth_verifier) {
        var requestTokenOauth = {
            consumer_key: TWITTER_CONSUMER_KEY,
            consumer_secret: TWITTER_CONSUMER_SECRET,
            callback: TWITTER_CALLBACK_URL
        };

        // Step 1. Obtain request token for the authorization popup.
        request.post({ url: requestTokenUrl, oauth: requestTokenOauth }, function(err, response, body) {
            var oauthToken = qs.parse(body);
            var params = qs.stringify({ oauth_token: oauthToken.oauth_token });

            // Step 2. Redirect to the authorization screen.
            res.redirect(authenticateUrl + '?' + params);
        });
    } else {
    var accessTokenOauth = {
        consumer_key: TWITTER_CONSUMER_KEY,
        consumer_secret: TWITTER_CONSUMER_SECRET,
        token: req.query.oauth_token,
        verifier: req.query.oauth_verifier
    };

    // Step 3. Exchange oauth token and oauth verifier for access token.
    request.post({ url: accessTokenUrl, oauth: accessTokenOauth }, function(err, response, profile) {
        profile = qs.parse(profile);

        var searchStr = '$filter=TwitterHandle eq \'@' + profile.screen_name + '\'';
        db2.getItem('users', searchStr, function(err, data) {
            if (err) {
                var msg = {
                    status: 'fail',
                    message: 'Error Retrieving User',
                    error: err
                };
                return res.send(JSON.stringify(msg));
            } else {
                if(req.headers.authorization && req.headers.authorization.indexOf('undefined') == -1) {
                    var token = req.headers.authorization.split(' ')[1];
                    var payload = jwt.decode(token, SESSION_SECRET);

                    if (payload.sub !== data[0].id) {
                        res.status(400).send({ message: 'User not found' });
                    } else {
                        res.send({ token: createToken(data[0]) });
                    }
                } else {
                    res.send({ token: createToken(data[0]) });
                }
            }
        });
    });
    }
});

/*****************
 * WINDOWS LIVE
 ****************/
app.post('/auth/live', function(req, res) {
    async.waterfall([
        function(done) {
            var accessTokenUrl = 'https://login.live.com/oauth20_token.srf';
            var params = {
                code: req.body.code,
                client_id: req.body.clientId,
                client_secret: WINDOWS_LIVE_CLIENT_SECRET,
                redirect_uri: req.body.redirectUri,
                grant_type: 'authorization_code'
            };
            request.post(accessTokenUrl, { form: params, json: true }, function(err, response, accessToken) {
                done(null, accessToken);
            });
        },
        function(accessToken, done) {
            var profileUrl = 'https://apis.live.net/v5.0/me?access_token=' + accessToken.access_token;
            request.get({ url: profileUrl, json: true }, function(err, response, profile) {
                done(err, profile);
            });
        },
        function(profile) {
            var email = profile.emails.account;
            var username;
            if (email) {
                username = email.match(/(\w+)@/)[1];
            }
            var searchStr = '$filter=Username eq \'' + username + '\'';
            db2.getItem('users', searchStr, function(err, data) {
                if (err) {
                    var msg = {
                        status: 'fail',
                        message: 'Error Retrieving User',
                        error: err
                    };
                    return res.send(JSON.stringify(msg));
                } else {
                    if(req.headers.authorization && req.headers.authorization.indexOf('undefined') == -1) {
                        var token = req.headers.authorization.split(' ')[1];
                        var payload = jwt.decode(token, SESSION_SECRET);

                        if (payload.sub !== data[0].id) {
                            res.status(400).send({ message: 'User not found' });
                        } else {
                            res.send({ token: createToken(data[0]) });
                        }
                    } else {
                        res.send({ token: createToken(data[0]) });
                    }
                }
            });
        }
    ]);
});

/*****************
 * FILE UPLOADS
 ****************/
app.post('/upload/img/author', upload.saveAuthorImage);
app.post('/upload/img/post', upload.savePostImage);

/*****************
 * PASSPORT (Deprecated)
 ****************/
var TWITTER_CALLBACK_URL_OLD       = c.get('TWITTER_CALLBACK_URL_OLD')
  , WINDOWS_LIVE_CALLBACK_URL_OLD  = c.get('WINDOWS_LIVE_CALLBACK_URL_OLD');

// Serialize users into sessions with just their user id
passport.serializeUser(function(user, cb){
    if (user) {
        cb(null, user.id);
    } else {
        cb(new Error('Couldn\'t serialize user'), null);
    }
});

passport.deserializeUser(db.deserializeUser);

passport.use(new TwitterStrategy({
    consumerKey: TWITTER_CONSUMER_KEY,
    consumerSecret: TWITTER_CONSUMER_SECRET,
    callbackURL: TWITTER_CALLBACK_URL_OLD
    }, function(token, tokenSecret, profile, done) {

    db.getUserFromTwitterProfile(profile, function(err, user){
        if (user && !err) {
            return done(null, user);
        } else {
            return err ? done(err, null) : done(new Error('User not authorized'), null);
        }

    });
}
));

passport.use(new LiveStrategy({
        clientID: WINDOWS_LIVE_CLIENT_ID,
        clientSecret: WINDOWS_LIVE_CLIENT_SECRET,
        callbackURL: WINDOWS_LIVE_CALLBACK_URL_OLD
    }, function(accessToken, refreshToken, profile, done) {
        console.log('access token: ' + accessToken);

        db.getUserFromLiveProfile(profile, function(err, user) {

        if (user && !err) {
            return done(null, user);
        } else {
            done('User not authorized', null);
        }

        });
    }
));

function auth (req, res, next) {
  if (req.isAuthenticated()) { return next(); }
  res.redirect('/admin-old/login');
};

function authAdmin (req, res, next) { // They are authenticated and authorized
  if (req.isAuthenticated() && req.user.IsAdmin === true) { return next(); }
  res.redirect('/admin-old/notadmin');
};

/*****************
 * TWITTER (Deprecated)
 ****************/
var message = 'Unable to authenticate, ' +
            'or you are not authorized to use this system.';

app.get('/auth-old/twitter',
  passport.authenticate('twitter'),
  function(req, res){/* isnt called */}
);

app.get('/auth-old/twitter/callback',
  passport.authenticate('twitter', {
    failureRedirect: '/admin-old/login?message=' + message
  }),
  function(req, res) {
    // Successful authentication, redirect home.
    res.redirect('/admin-old/post');
});

/*****************
 * WINDOWS LIVE (Deprecated)
 ****************/
app.get('/auth-old/live',
  passport.authenticate('windowslive', { scope: ['wl.signin', 'wl.basic', 'wl.emails'] }));

app.get('/auth-old/live/callback',
  passport.authenticate('windowslive', {
    failureRedirect: '/admin-old/login?message=' + message
   }),
  function(req, res) {
    // Successful authentication, redirect home.
    res.redirect('/admin-old/post');
  });

/*****************
 * Old Site Redirects
 ****************/
app.get('/blogs/post/*', redirects);
app.get('/analytics', function(req, res) { res.redirect('/post/120085') });

/*****************
 * 404 Redirect
 ****************/
// TODO: this the wrong way to send a 404, it should be
// `res.status(404).render('404');` but that doesn't render correctly on Azure.
app.get('/*', function(req, res) { res.render('404'); });

/**********************************
 * START THE SERVER, SCOTTY!
 **********************************/
app.listen(process.env.PORT || port);
console.log('Express server listening on port ' + port);
